inference-engine/src/vpu/common/src/ngraph/transformations/merge_gather_gather_elements.cpp

```diff
@@ -0,0 +1,156 @@
+    const auto squeeze = ngraph::pattern::wrap_type<ngraph::opset6::Squeeze>({gather, squeezeAxes}, ngraph::pattern::consumers_count(1));
+
+    const auto transposePerm = ngraph::pattern::any_input(
+        ngraph::pattern::op::as_value_predicate(ngraph::pattern::has_class<ngraph::opset6::Constant>()));
+    const auto transpose = ngraph::pattern::wrap_type<ngraph::opset6::Transpose>({squeeze, transposePerm});
+
+    const auto m = std::make_shared<ngraph::pattern::Matcher>(transpose, "GatherSqueezeTransposeMatcher");
+    for (const auto& node : f->get_ordered_ops()) {
+        if (!m->match(node)) {
+            continue;
+        }
+
+        auto& patternMap = m->get_pattern_value_map();
+
+        std::vector<ngraph::Node*> shapeOfs;
+        std::vector<ngraph::Node*> gatherElements;
+
+        const auto& matchedTranspose = patternMap.at(transpose);
+        const auto& transposeConsumers = matchedTranspose.get_target_inputs();
+
+        const auto isShapeOfOrGatherElements = [](const ngraph::Input<ngraph::Node>& consumer) {
+            return consumer.get_node()->get_type_info() == ngraph::opset6::ShapeOf::type_info ||
+                   consumer.get_node()->get_type_info() == ngraph::opset6::GatherElements::type_info;
+        };
+
+        if (!std::all_of(transposeConsumers.cbegin(), transposeConsumers.cend(), isShapeOfOrGatherElements)) {
+            continue;
+        }
+
+        for (const auto& transposeConsumer : transposeConsumers) {
+            if (transposeConsumer.get_node()->get_type_info() == ngraph::opset6::ShapeOf::type_info) {
+                shapeOfs.push_back(transposeConsumer.get_node());
+            } else if (transposeConsumer.get_node()->get_type_info() == ngraph::opset6::GatherElements::type_info) {
+                gatherElements.push_back(transposeConsumer.get_node());
+            }
+        }
+
+        const auto& matchedTransposePerm = patternMap.at(transposePerm);
+        const auto& matchedSqueezeAxes = patternMap.at(squeezeAxes);
+        const auto& matchedSqueeze = patternMap.at(squeeze);
+        const auto& matchedGatherData = patternMap.at(gatherData);
+        const auto& matchedGatherIndices = patternMap.at(gatherIndices);
+        const auto& matchedGather = patternMap.at(gather);
+
+        for (const auto& gatherElement : gatherElements) {
+            const auto transposeIndices = std::make_shared<ngraph::opset6::Transpose>(gatherElement->input_value(1), matchedTransposePerm);
+            const auto unsqueezeIndices = std::make_shared<ngraph::opset6::Unsqueeze>(
+                transposeIndices,
+                matchedSqueezeAxes);
+            const auto expGatherElements = std::make_shared<ngraph::vpu::op::ExpGatherElements>(
+                matchedGatherData,
+                unsqueezeIndices,
+                matchedGatherIndices,
+                ngraph::as_type<ngraph::opset6::GatherElements>(gatherElement)->get_axis(),
```

Comment:
`get_exis` will return max int in case if it can not extract single constant from axis input. 
Also I suggest to connect axis inputs from Gather and GatherElements to ExpGatherElements. And in ExpGatherElements validate_and_infer you can use `get_constant_from_source` to get Constant from particular input.

Reply:
In case of `GatherElements` `axis` is a simple attribute, so there is no other way to do so, and I did the same for `ExpGatherElements` - both axes are attributes.
Fixed this for `Gather`'s axis.